web/book/src/internals/modules.md

```diff
@@ -0,0 +1,325 @@
+    # _my_project.prql
+
+    module sales {
+        # sales.prql
+    }
+
+    module projections {
+        # projections/_projections.prql
+
+        module year_2023 {
+            # projections/year_2023.prql
+        }
+
+        module year_2024 {
+            # projections/year_2024.prql
+        }
+    }
+}
+```
+
+If module `my_project.sales` wants to add a submodule `util`, it has to be
+converted to a directory modules. This means that it has to be moved to
+`sales/_sales.prql`. The new module would reside in `sales/util.prql`.
+
+The annotated layout is not the only possible layout for this module structure,
+since any of the modules `sales`, `year_2023` or `year_2024` could be converted
+into a directory module with zero sub-modules.
+
+Point 4 of design goals means that each declaration within a project has a
+single fully-qualified name within this project. This is ensured by strict rules
+regarding importing files and the fact that the module structure is a tree.
+
+## Declaration order
+
+The order of declarations in a module holds no semantic value, except the "last
+`main` variable".
+
+References between modules can be cyclic.
+
+```
+module mod_a {
+    let decl_a_1 = ...
+    let decl_a_2 = (from mod_b.decl_b | take 10)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a | take 10)
+}
+```
+
+References between variable declarations cannot be cyclic.
+
+```
+let decl_a = (from decl_b)
+let decl_b = (from decl_a) # error: cyclic reference
+```
+
+```
+module mod_a {
+    let decl_a = (from mod_b.decl_b)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a) # error: cyclic reference
+}
+```
+
+## Compiler interface
+
+`prql-compiler` provides two interfaces for compiling files.
+
+**Multi-file interface** requires three arguments:
+
+- path to the file containing the module which is the root of the module
+  structure,
+- identifier of the pipeline that should be compiled to RQ (this can also be an
+  identifier of a module that has a `main` pipeline) and,
+- a "file loader", which can load files on-demand.
+
+The path to the root module can be automatically detected by searching for
+`.prql` files starting with `_` in the current working directory.
+
+Example prqlc usage:
+
+```
+$ prqlc compile _project.prql sales.projections.orders_2024
+$ prqlc compile sales.projections.orders_2024
+```
+
+**Single-file interface** requires a single argument; the PRQL source. Any
+attempts to load modules in this mode result in compilation errors. This
+interface is needed, for example, when integrating the compiler with a database
+connector (i.e. JDBC) where no other files can be loaded.
+
+## Built-in module structure
+
+> Work In Progress
+
+```
+# root module of every project
+module project {
+	module std {
+		func sum a -> ...
+		func mean a -> ...
+	}
+
+	module default_db {
+		# all inferred tables and defined CTEs
+	}
+
+	let main = (
+		from t = tracks
+		select [track_id, title]
+	)
+}
+```
+
+## Example
+
```

Comment:
@max-sixty Another thread for "Opaqueness of idents".

----

(Bear with me, I need to explain all this to so I can explain how I'd change it)

Right now, when we infer that a table exists, we add its declaration to module `default_db` (this declaration is updated when we discover new columns).

Using a made-up syntax:
```
module default_db {
    table foo = ...
}
```

If a table name is segmented (i.e. `my_db.my_schema.foo`), we infer this:
```
module default_db {
    table `my_db.my_schema.foo` = ...
}
```
... essentially, the table name contains the dots, which is then problematic when actual dots are inserted into the name segments (i.e. ``my_db.`my.schema`.foo``).

----

Instead, I propose we infer this:
```
module default_db {
    module my_db {
        module my_schema {
            table foo = ...
        }
    }
}
```

Reply:
Yes, very much agree. that's what I'd been trying to do in my branch for #1535 — when it finds a table type of `my_schema.foo`, it would create a module `my_schema` in the root (though should actually be in `default_db`, good point).

My branch is super-messy, but maybe I push it and you can criticize / guide it towards something helpful. I haven't worked on it in a couple of weeks now.

Reply:
That's a lot of indention. Maybe it could be:

    module default_db.my_db.my_schema {
        table foo = ... # just one level of indention
    }
    
Or just remove the indention all together and make one module per file:

    module default_db.my_db.my_schema;
    
    table foo = ... # no indention

But why does the module even need to be explicitly declared? The module could be the file, then the filename is the module name.

Reply:
Yes, that makes sense. My bad, I had misunderstood it.